File: kamerider_image_detection/src/person_detection.cpp
//找到一个人并走到人跟前。通过人脸的大小，认为到了人跟前。   
#include <ros/ros.h>    
//Use image_transport for publishing and subscribing to images in ROS    
#include <image_transport/image_transport.h>    
//Use cv_bridge to convert between ROS and OpenCV Image formats    
#include <cv_bridge/cv_bridge.h>    

#include <sensor_msgs/image_encodings.h>    
//Include headers for OpenCV Image processing    
#include <opencv2/imgproc/imgproc.hpp>    
//Include headers for OpenCV GUI handling    
#include <opencv2/highgui/highgui.hpp>    
#include <string>        
#include <sstream>    
#include <stdio.h>
#include <opencv2/opencv.hpp>
#include <iostream>
#include <fstream>
#include <math.h>
#include <std_msgs/String.h>
#include <geometry_msgs/Twist.h>
#include <vector>
#include <stdlib.h>
#include <kamerider_image_msgs/mission.h>
int system(const char *string); 
using namespace cv;    
using namespace std;    
int tour=0;
//Store all constants for image encodings in the enc namespace to be used later.      
namespace enc = sensor_msgs::image_encodings; 
struct PathPoint
{
    int state;
    float vel;
    float time;
};
vector<PathPoint> vppt;             //储存反馈调节时的调节参数
void image_process(Mat inImg); 
void spCallback(const std_msgs::String::ConstPtr& msg); 
void forward_robot(float x);
void turn_robot(float theta);
/// 全局变量

const char* window_name = "person_detection";
String face_cascade_name = "/home/nvidia/catkin_ws/src/kamerider_image/kamerider_image_detection/dataset/haarcascade_fullbody.xml";  
//String eyes_cascade_name = "/home/isi/2017_ws/src/imgpcl/haarcascade_eye_tree_eyeglasses.xml";  
CascadeClassifier face_cascade;

ros::Publisher move_pub; //向navigation发消息，微调机器人
ros::Publisher result_pub;   //检测到物体的数量
ros::Publisher image_pub;
ros::Subscriber control_sub;
ros::Subscriber image_sub;

std_msgs::String msg_to_control;
//ros::Publisher iffind_person;
std_msgs::String iffind_personP;
int state = -1;
//This function is called everytime a new image is published    
void imageCallback(const sensor_msgs::ImageConstPtr& original_image)    
{

    //Convert from the ROS image message to a CvImage suitable for working with OpenCV for processing    
    cv_bridge::CvImagePtr cv_ptr;      
    try      
    {      
        //Always copy, returning a mutable CvImage      
        //OpenCV expects color images to use BGR channel order.      
        cv_ptr = cv_bridge::toCvCopy(original_image, enc::BGR8);      
    }      
    catch (cv_bridge::Exception& e) 
    {      
        //if there is an error during conversion, display it      
        ROS_ERROR("tutorialROSOpenCV::main.cpp::cv_bridge exception: %s", e.what());      
        return;
    }      
    image_process(cv_ptr->image);    
}
void image_process(Mat inImg)    
{
    if(state == -1 || state == 2)
        return;
    if( inImg.empty() )    
    {    
        ROS_INFO("Camera image empty");    
        return;//break;    
    } 
    if( !face_cascade.load( face_cascade_name ) ){ printf("--(!)Error loading\n"); };  
    //if( !eyes_cascade.load( eyes_cascade_name ) ){ printf("--(!)Error loading\n"); };
    std::vector<Rect> faces;  
    Mat inImg_gray;  
    cvtColor( inImg, inImg_gray, CV_BGR2GRAY );  
    equalizeHist( inImg_gray, inImg_gray );  
    //-- Detect faces  
    face_cascade.detectMultiScale( inImg_gray, faces, 1.1, 3, 0|CV_HAAR_SCALE_IMAGE, Size(30, 30) );  
    int human_num = faces.size();
    cout<<"human_num:"<<human_num<<endl;
    if(human_num == 0)
    {
        //msg_to_control.data = "I find no people, I am finding";
        //image_pub.publish(msg_to_control);//发送此时检测到的物体数量
        turn_robot(1.5);
	tour+=1.5;
	if(tour>=6.28)
	{
		tour-=6.28;
		msg_to_control.data=" please get away from me ";
		image_pub.publish(msg_to_control);
	}
    }
    if(human_num >= 1)
    {   
        //state = 2;
        //iffind_personP.data = "found_person";
        //result_pub.publish(iffind_personP); 
        msg_to_control.data = "I am reaching the person";
        image_pub.publish(msg_to_control);
        Mat face = inImg_gray( faces[0] ); 
   
        rectangle(inImg,faces[0],Scalar(0,255,0),1);
        //imwrite("/home/isi/2017_ws/src/imgpcl/person.jpg",inImg);
        int pos_x = 0;
        int pos_y = 0;
        pos_x = faces[0].x + faces[0].width ;
        pos_y = faces[0].y + faces[0].height;
        cout<<"pos_x"<<pos_x<<endl;
        cout<<"pos_y"<<pos_y<<endl;
        if(pos_x - 320 > 80)
            turn_robot(-0.8);
        if(pos_x - 320 < -80)
            turn_robot(0.8);
        if(pos_y > 280)
            forward_robot(0.5);
        if(abs(pos_x - 320) < 80 && pos_y < 280)
	    {
	        cout<<"OK"<<endl;
            msg_to_control.data = "I have reached the person!";
            image_pub.publish(msg_to_control); 
            iffind_personP.data = "found_person";
            result_pub.publish(iffind_personP);                       		    state = 2;		
	    }
    }
	imwrite("/home/nvidia/catkin_ws/src/kamerider_image/kamerider_image_detection/result/found_person.jpg",inImg);//保存图片    
    imshow( window_name, inImg );
    waitKey(5); 
} 
void turn_robot(float theta)
{
    geometry_msgs::Twist vel;
    int count = 0;
    float time = 2;
    ros::Rate loop_rate(10);
    int num = time*10;
    PathPoint pt;
    //转90度
    //float theta = 5;
	//cout<<"theta:"<<theta<<endl;
	float theta2 = theta/time;
	vel.angular.z = theta2;    
    count = 0;
    num = time*10;
    while(count < num)
    {
        count++;
        move_pub.publish(vel);
        loop_rate.sleep();
    }
    pt.state = 0;
    pt.time = time;
    pt.vel = vel.angular.z;
    vppt.push_back(pt);
    vel.angular.z = 0.0;
    move_pub.publish(vel);
    cout<<"转弯\n";
    sleep(1);
}
void forward_robot(float x)
{
    geometry_msgs::Twist vel;
    int count = 0;
    float time = 4;
    vel.linear.x = x/time; 
    ros::Rate loop_rate(10);
    int num = time*10;
    //cout<<"armCallback1()\n";
    while(count < num)
    {
        count++;
        move_pub.publish(vel);
        loop_rate.sleep();
    }
    PathPoint pt;
    pt.state = 0;
    pt.time = time;
    pt.vel = vel.linear.x;
    vppt.push_back(pt);
    vel.linear.x = 0.0;
    move_pub.publish(vel);
    cout<<"直行\n";
    sleep(1);
}
void ctrlCallback(const kamerider_image_msgs::mission::ConstPtr& msg)
{
    if (msg->mission_type == "person")
    {
        //cout<<"ctrlCallback()\n";
        ROS_INFO("ctrlCallback received: %s\n", msg->mission_name.c_str());
        state = 0;
    }
}   
/** 
* This is ROS node to track the destination image 
*/    
int main(int argc, char **argv)    
{    
    ros::init(argc, argv, "person_detection");    
    ROS_INFO("-----------------");
    ros::NodeHandle nh;
    move_pub = nh.advertise<geometry_msgs::Twist>("cmd_vel_mux/input/navi",1);  //移动
    result_pub = nh.advertise<std_msgs::String>("/found_person", 1);
    image_pub = nh.advertise<std_msgs::String>("/kamerider_speech/input", 1);

    control_sub = nh.subscribe("/control_to_image", 1, ctrlCallback);
    image_sub = nh.subscribe("/astra/rgb/image_raw", 1, imageCallback);
    image_transport::ImageTransport it(nh); 
   // ros::Subscriber camInfo         = nh.subscribe("camera/rgb/camera_info", 1, camInfoCallback);    
    // ros::Subscriber arm_sub = nh.subscribe("/control_to_image", 1, ctrlCallback);        //订阅navigation的消息
    ros::spin();
    //ROS_INFO is the replacement for printf/cout.    
    ROS_INFO("main_No error.");  
}

File: kamerider_image_detection/src/person_detection.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
Date: 2019/2/1
Author: Xu Yucheng
Abstract: Code for person detection
'''
import os
import cv2
import math
import rospy
import roslib
import base64
from aip import AipBodyAnalysis
from std_msgs.msg import Int8
import matplotlib.pyplot as plt
from std_msgs.msg import String
from sensor_msgs.msg import Image
from geometry_msgs.msg import Twist
from cv_bridge import CvBridge, CvBridgeError
from kamerider_image_msgs.msg import mission



class person_detection:
    def __init__(self):
        # 定义baidu ai 的APP_ID、API_KEY、SECRET_KEY
        """
        根据认证信息获取access_token.
        access_token的有效期为30天，切记需要每30天进行定期更换，或者每次请求都拉取新token；
        """
        self.APP_ID     = '16002184'
        self.API_KEY    = 'W6jgTXcHFZioPcmXWilfaGO2'
        self.SECRET_KEY = 'oNewRv2mtqfknwzq3rgEFc3sP4SAbI0v'
        
        #ros params
        self.msg_count = 0
        self._turn = False
        self.find_person = True
        self.start_detect = False
        self.angle_count = 0
        self.take_photo_signal = False
        self.sub_image_raw_topic_name=None
        self.sub_control_topic_name = None
        self.pub_person_detect_topic_name=None

        self.path_to_save_image=None
        self.pub_result=None
        self.move_pub = None
        self.speech_pub = None
        self.client = None
        self.get_params()
        self.turn_robot(1.8)
        while (True):
            if self.find_person == False:
                rospy.loginfo("Msg received start person detection")
                if self.start_detect == True:
                    self.detection()

    def get_params(self):
        self.sub_image_raw_topic_name          = rospy.get_param('sub_image_raw_topic_name',          '/astra/rgb/image_raw')
        self.sub_control_topic_name            = rospy.get_param("sub_control_topic_name",            '/control_to_image')
        self.pub_person_detect_topic_name      = rospy.get_param('pub_person_detect_topic_name',      '/image_to_control')
        self.pub_to_move_base_topic_name       = rospy.get_param('pub_to_move_base_topic_name',       '/cmd_vel_mux/input/navi')
        self.pub_to_speech_topic_name          = rospy.get_param("pub_to_speech_topic_name",          '/kamerider_speech/input')
        self.path_to_save_image                = rospy.get_param('path_to_save_image',                '/home/nvidia/catkin_ws/src/kamerider_image/kamerider_image_detection/result/person_image_capture.jpg')
        self.path_to_save_result               = rospy.get_param('path_to_save_result',               '/home/nvidia/catkin_ws/src/kamerider_image/kamerider_image_detection/result/person_detection_result.jpg')     
        #定义R发布器和订阅器，话题名通过ROS PARAM参数服务器获取
        rospy.Subscriber(self.sub_image_raw_topic_name, Image, self.imageCallback)
        rospy.Subscriber(self.sub_control_topic_name, mission, self.controlCallback)
        self.pub_result = rospy.Publisher(self.pub_person_detect_topic_name, String, queue_size=1)
        self.move_pub   = rospy.Publisher(self.pub_to_move_base_topic_name, Twist, queue_size=1)
        self.speech_pub = rospy.Publisher(self.pub_to_speech_topic_name, String, queue_size=1)
    
    def controlCallback(self, msg):
        if msg.mission_type == "person":
            self.msg_count += 1
            if self.msg_count > 1:
                self.forward_robot(-0.5)
                self.turn_robot(1.8)
            self.find_person = False
            self.take_photo_signal = True

    def imageCallback(self, msg):
        bridge = CvBridge()
        cv_image = bridge.imgmsg_to_cv2(msg, 'bgr8')
        #cv2.imshow("current", cv_image)
        #cv2.waitKey(10)
        if self.take_photo_signal == True:
            print ("[INFO] Start to take photo")
            try:
                cv2.imwrite(self.path_to_save_image, cv_image)
                cv2.imshow("frame", cv_image)
                cv2.waitKey(100)
            except CvBridgeError as e:
                print (e)
            self.start_detect = True
            self.take_photo_signal = False
            
    def image_encode(self):
        with open(self.path_to_save_image, 'rb') as f:
            return f.read()

    #检测函数
    def detection(self):
        self.start_detect = False
        encode_image = self.image_encode()
        print (type(encode_image))
        image = cv2.imread(self.path_to_save_image)
        #初始化aipfacce对象
        client = AipBodyAnalysis(self.APP_ID, self.API_KEY, self.SECRET_KEY)
        #接收返回的检测结果
        result = client.bodyAttr(encode_image)
        print (result)
        num = result["person_num"]
        if result["person_num"] == 0:
            rospy.loginfo("Start Turning robot")
            self.turn_robot(1.5)
            self.angle_count += 1.5
            self.take_photo_signal = True

        if result["person_num"] > 1:
            self.speech_pub.publish("I can see more than one person, I have found the target person")
            self.speech_pub.publish("Please stand in front of me and lead me")
            for i in range(num):
                person_info = result["person_info"][i]
                point1 = (person_info["location"]["left"], person_info["location"]["top"])
                point2 = (person_info["location"]["left"]+person_info["location"]["width"], person_info["location"]["top"]+person_info["location"]["height"])
                cv2.rectangle(image, point1, point2, (0,0,255), 2)
            cv2.imwrite(self.path_to_save_result, image)
            self.find_person = True
            msg = String()
            msg.data = "person_target_found"
            self.pub_result.publish(msg)

        if result["person_num"] == 1:
            person_info = result["person_info"][0]
            person_x = person_info["location"]["left"] + person_info["location"]["width"]/2
            person_y = person_info["location"]["top"] + person_info["location"]["height"]/2
            point1 = (person_info["location"]["left"], person_info["location"]["top"])
            point2 = (person_info["location"]["left"]+person_info["location"]["width"], person_info["location"]["top"]+person_info["location"]["height"])
            
            cv2.rectangle(image, point1, point2, (0,0,255), 2)
            cv2.imwrite(self.path_to_save_result, image)
            
            if person_x-320 > 80:
                self.turn_robot(-1.1)
            if person_x -320 < -80:
                self.turn_robot(1.1)
            if (person_info["location"]["top"] < 240):
                self.forward_robot(0.5)
                self.find_person = True
                rospy.loginfo("Person Target Reached!")
                control_msg = String()
                speech_msg = String()
                control_msg.data = "person_target_found"
                speech_msg.data = "I have reached the target person"
                self.speech_pub.publish(speech_msg)
                self.pub_result.publish(control_msg) 
        
            
    def turn_robot(self, angle):
        vel = Twist()
        count = 0
        time = 2
        rate = 10
        r = rospy.Rate(rate)
        num = time*10
        theta = angle/time
        vel.angular.z = theta
        while (count < num):
            count += 1
            self.move_pub.publish(vel)
            r.sleep()
        vel.angular.z = 0.0
        self.move_pub.publish(vel)
        print ("Turn robot~")
    
    def forward_robot(self, dis):
        vel = Twist()
        count = 0
        time = 4
        vel.linear.x = dis/time
        rate = 10
        r = rospy.Rate(rate)
        num = time*10
        while (count < num):
            count += 1
            self.move_pub.publish(vel)
            r.sleep()
        
        vel.linear.x = 0.0
        self.move_pub.publish(vel)
        print ("Go forward~")
        

if __name__ == '__main__':
    #初始化节点
    rospy.init_node('person_detection')
    print ('----------init----------')
    print ('-----WAITING FOR IMAGE-----')
    detector = person_detection()
    rospy.spin()
        


